File: Algorithms & Data Structures/semester3/graphs/b1.cpp
#include <bits/stdc++.h>

using namespace std;

const long long maxn = 2e5;
vector<pair<long long, long long>> g[maxn];
long long d[maxn];
bool u[maxn];
vector<long long> top;

void dfs(long long v) {
    u[v] = 1;
    for (auto [to, _]: g[v]) {
        if (!u[to]) {
            dfs(to);
        }
    }
    top.push_back(v);
}

signed main() {
    long long n, m;
    long long start, t;
    cin >> n >> m >> start >> t;
    for (long long i = 0, q, w, e; i < m; ++i) {
        cin >> q >> w >> e;
        g[q - 1].emplace_back(w - 1, e);
    }

    for (long long i = 0; i < n; ++i) {
        if (!u[i]) {
            dfs(i);
        }
    }

    reverse(top.begin(),top.end());

    fill(d, d + n, 1e18);
    d[start - 1] = 0;

    long long ind = find(top.begin(),top.end(), start - 1) - top.begin();

    for (long long i = ind; i < n; ++i) {
        long long v = top[i];
        for (auto [to, dist]: g[v]) {
            if (d[to] > d[v] + dist) {
                d[to] = d[v] + dist;
            }
        }
    }
    if (d[t - 1] >= 1e9) {
        cout << "Unreachable";
    }
    else {
        cout << d[t - 1];
    }
}

File: Algorithms & Data Structures/semester3/graphs/b1.rs
const N: usize = 200001;
const E_18: i64 = 1_000_000_000_000_000_000;
const E_9: i64 = 1_000_000_000;

fn dfs(
    vertex: usize,
    graph: &Vec<Vec<(usize, i64)>>,
    topological: &mut Vec<usize>,
    used: &mut Vec<bool>,
) {
    used[vertex] = true;
    for (to, _) in graph[vertex].iter() {
        if !used[*to] {
            dfs(*to, graph, topological, used);
        }
    }

    topological.push(vertex);
}

pub fn main() {
    let mut buffer = String::new();
    std::io::stdin()
        .read_line(&mut buffer)
        .expect("Failed io input!");

    let nmst = buffer
        .split_whitespace()
        .map(|s| s.parse::<usize>().unwrap())
        .collect::<Vec<_>>();

    let n = nmst[0];
    let m = nmst[1];
    let s = nmst[2];
    let t = nmst[3];

    let mut graph = vec![Vec::new(); N];
    let mut answer = vec![0; N];
    let mut used = vec![false; N];
    let mut topological = vec![];

    for _ in 0..m {
        buffer.clear();
        std::io::stdin()
            .read_line(&mut buffer)
            .expect("Failed to read line!");
        let uvw = buffer
            .split_whitespace()
            .map(|s| s.parse::<i64>().unwrap())
            .collect::<Vec<_>>();

        let u = uvw[0] as usize;
        let v = uvw[1] as usize;
        let w = uvw[2];

        graph[u - 1].push((v - 1, w));
    }

    for i in 0..n {
        if !used[i] {
            dfs(i, &graph, &mut topological, &mut used);
        }
    }

    topological.reverse();

    answer.fill(E_18);
    answer[s - 1] = 0;

    let start_idx = topological.iter().position(|&x| x == s - 1).unwrap();

    for i in start_idx..n {
        let v = topological[i];
        for (to, dist) in graph[v].iter() {
            if answer[*to] > answer[v] + dist {
                answer[*to] = answer[v] + dist;
            }
        }
    }

    if answer[t - 1] >= E_9 {
        println!("Unreachable");
    } else {
        println!("{}", answer[t - 1]);
    }
}
